x-pack/auditbeat/module/system/login/utmp.go

```diff
@@ -0,0 +1,571 @@
+	}
+
+	switch utmp.UtType {
+	// See utmp(5) for C constants.
+	case C.RUN_LVL: // 1
+		// The runlevel - though a number - is stored as
+		// the ASCII character of that number.
+		runlevel := string(rune(utmp.UtPid))
+
+		// 0 - halt; 6 - reboot
+		if utmp.UtUser == "shutdown" || runlevel == "0" || runlevel == "6" {
+			record.Type = LoginRecordTypeShutdown
+
+			// Clear any old logins
+			// TODO: Issue logout events for login events that are still around
+			// at this point.
+			r.loginSessions = make(map[string]LoginRecord)
+		} else {
+			// Ignore runlevel changes that are not halt or reboot.
+			return nil
+		}
+	case C.BOOT_TIME: // 2
+		if utmp.UtLine == "~" && utmp.UtUser == "reboot" {
+			record.Type = LoginRecordTypeBoot
+
+			// Clear any old logins
+			// TODO: Issue logout events for login events that are still around
+			// at this point.
+			r.loginSessions = make(map[string]LoginRecord)
+		} else {
+			// Ignore unknown record
+			return nil
+		}
+	case C.USER_PROCESS: // 7
+		record.Type = LoginRecordTypeUserLogin
+
+		record.Username = utmp.UtUser
+		record.UID = lookupUsername(record.Username)
+		record.PID = utmp.UtPid
+		record.IP = newIP(utmp.UtAddrV6)
+		record.Hostname = utmp.UtHost
+
+		// Store TTY from user login record for enrichment when user logout
+		// record comes along (which, alas, does not contain the username).
+		r.loginSessions[record.TTY] = record
+	case C.DEAD_PROCESS: // 8
+		savedRecord, found := r.loginSessions[record.TTY]
+		if found {
+			record.Type = LoginRecordTypeUserLogout
+			record.Username = savedRecord.Username
+			record.UID = savedRecord.UID
+			record.PID = savedRecord.PID
+			record.IP = savedRecord.IP
+			record.Hostname = savedRecord.Hostname
+		} else {
+			// Skip - this is usually the DEAD_PROCESS event for
+			// a previous INIT_PROCESS or LOGIN_PROCESS event -
+			// those are ignored - (see default case below).
+			return nil
+		}
+	default:
+		/*
+			Every other record type is ignored:
+			- EMPTY - empty record
+			- NEW_TIME and OLD_TIME - could be useful, but not written when time changes,
+			  at least not using `date`
+			- INIT_PROCESS and LOGIN_PROCESS - written on boot but do not contain any
+			  interesting information
+			- ACCOUNTING - not implemented according to manpage
+		*/
+		return nil
+	}
+
+	return &record
+}
+
+// lookupUsername looks up a username and returns its UID.
+// It does not pass through errors (e.g. when the user is not found)
+// but will return -1 instead.
+func lookupUsername(username string) int {
+	if username != "" {
+		user, err := user.Lookup(username)
+		if err == nil {
+			uid, err := strconv.Atoi(user.Uid)
+			if err == nil {
+				return uid
+			}
+		}
+	}
+
+	return -1
+}
+
+func newIP(utAddrV6 [4]uint32) *net.IP {
+	var ip net.IP
+
+	// See utmp(5) for the utmp struct fields.
+	if utAddrV6[1] != 0 || utAddrV6[2] != 0 || utAddrV6[3] != 0 {
+		// IPv6
+		b := make([]byte, 16)
+		binary.LittleEndian.PutUint32(b[:4], utAddrV6[0])
```

Comment:
What does this mean for big-endian architectures? Will this code work?

Reply:
Hm, good point and I don't know. Do we support any big-endian systems?

Reply:
I'm surprised Go doesn't have a binary.MachineEndian.

Reply:
```
package main

import (
	"encoding/binary"
	"fmt"
	"unsafe"
)

var MachineEndian = getByteOrder()

func getByteOrder() binary.ByteOrder {
	var b [2]byte
	*((*uint16)(unsafe.Pointer(&b[0]))) = 1
	if b[0] == 1 {
		return binary.LittleEndian
	}
	return binary.BigEndian
}

func main() {
	var b [4]byte
	MachineEndian.PutUint32(b[:], 0x12345678)
	fmt.Printf("%x %x %x %x\n", b[0], b[1], b[2], b[3])
}

```

Reply:
Just revisiting this. I'm thinking that it might not be likely that this would be used on a big-endian system that uses UTMP login records. In the interest of keeping the code simple I'd leave it as is. If there's a need to change things we always can.

@adriansr `getByteOrder()` looks like a useful function that Go itself should have, indeed. Or maybe we should have it in a shared location within Beats so code like this here can use it easily.

Reply:
I think it's likely that someone will use it on big-endian. We've had several contributions to go-libaudit to make auditbeat work on big-endian.

Reply:
Ok, I've added support for it. Thanks @adriansr for the code snippet!